admin/server.go

```diff
@@ -0,0 +1,57 @@
+package admin
+
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"github.com/reddit/baseplate.go/log"
+)
+
+const (
+	DefaultAdminAddr = ":6060"
+)
+
+// ServerArgs contain optional configuration for the admin server.
+type ServerArgs struct {
+	// AdminAddr is a custom address for the admin server. Defaults to port 6060 on localhost if not set.
+	AdminAddr string
+	// HealthCheckFn is the HTTP handler for health check.
+	HealthCheckFn http.HandlerFunc
+}
+
+// NewServer returns a new admin server for internal functionality.
+func NewServer(args *ServerArgs) *Server {
+	adminAddr := args.AdminAddr
+	if args.AdminAddr == "" {
+		adminAddr = DefaultAdminAddr
+	}
+
+	return &Server{
+		adminAddr:     adminAddr,
+		healthCheckFn: args.HealthCheckFn,
+	}
+}
+
+type Server struct {
+	adminAddr     string
+	healthCheckFn http.HandlerFunc
+}
```

Comment:
I find the names `NewServer` and `Server` too broad, like they can only handle admin requests but the name seem to imply they are the main server. I'd rather just have `func ServeAdmin(cfg AdminConfig)` and call it a day.

Reply:
I am in agreement about the top-level function part, but I have experience with this kind of API and the struct options ended up in pain.  We had to pivot to variadic options to stay sane.  That said, for now I think being strongly opinionated will end up better for us, since we have a small number of entrypoints.

I think we should go for
```
package httpbp
func ServeAdmin(isHealthy http.Handler)

// and

package thriftbp
func ServeAdmin()
```

to enforce the strong opinions we want.

Reply:
@fishy re: `NewServer` being too broad ... the entire call with package name will be `admin.NewServer()` so that is why I originally removed the `admin` from `ServeAdmin` because it seemed redundant like `admin.ServeAdmin`. I'm fine to do it either way, I don't feel strongly one way or the other.

@kylelemons  I'm interested to understand how this "kind of API and the struct opts end up a pain"?  And Im interested to understand why variadic options was an improvement?  Re: "be strongly opinionated" ... I was trying to avoid being strongly opinionated (i.e. passing `isHealthy` as an arg) since it seems it will make backwards compatibility harder in the future having a strict API.

Reply:
Struct options are one way to handle optional parameters, but they are optimized for cases where _every_ call site has to pass some parameters because of the added syntax and cognitive overhead.  Compared to the alternatives, they're best when every call site needs to pass 2+ options.  If most call-sites need zero or one option, if option values require validation, if you have multiple use-cases, if options themselves have options, or if you want the flexibility to adjust internals, the variadic option pattern is a good choice.  If you only have a few use-cases, having specific top-level functions with non-optional parameters is even better still.

For the hole we fell into with an API like this specifically: we started out with just a "Port" field, but that turned out to be a pain to test because in tests you want to use a random port.  So we allowed passing in zero, but then we needed a way to exfiltrate the port that was chosen.  So we ended up also having a "Listener" field, but now you have this weird "what if both the listener AND the port are provided?" question.  I think similar things also happened around logging and error handling and the mux and some other configuration.  Using variadic options simplified these a lot, for example `admin.Serve(admin.WithPort(123))` vs `admin.Serve(admin.WithListener(lis))` and now you can also have `admin.Serve(admin.WithAddr("localhost:1234")` without having to have _three_ possibly-conflicting fields.

As I mentioned elsewhere though, having top-level APIs that are opinionated and purpose-built is often the. best overall.  `thrfitbp.ServeAdmin()` + `httpbp.ServeAdmin(isHealthy)` for example -- with this arrangement, there is no "optional" arguments to worry about.  We can still expose the "low level" `admin` library, but caution that most people should not use it directly.

Reply:
Interesting. Thanks for the extra info @kylelemons.  
So if we do `thrfitbp.ServeAdmin()` + `httpbp.ServeAdmin(isHealthy)` which then removes the need for the `AdminServerArgs struct`, then we should not allow an adminAddr as an option? Or do we want `addr` to still be configurable and in that case have:
`thrfitbp.ServeAdmin(addr)` + `httpbp.ServeAdmin(addr, isHealthy)`
?

Trying to remove the `AdminServerArgs struct` for now seems ok, but what about down the road if we end up wanting to add more options... following this pattern we could end up with more servers like ... 
```
thrfitbp.ServeAdmin(addr) 
thrfitbp.ServeAdmin(addr, newNeededOpt) 
httpbp.ServeAdmin(addr, isHealthy)
httpbp.ServeAdmin(addr, isHealthy, newNeededOpt)
httpbp.ServeAdmin(addr, newNeededOpt)
```
Is that right? Or am I misunderstanding something?

Reply:
Listening on any address other than 0.0.0.0 is basically incorrect for kubernetes, so no I don't think providing the address is something we need.

I can't really come up with any additional configuration or customization that I think we would want here.  We have the option of exposing an admin package in the future to allow for deeper customization but I'm not really sure what we would want to let people tweak.  If we did have them though, we could add `http.ServeAdminOpts` or we could invoke the clause in the baseplate 2022 compatibility promise that "if it doesn't break any visible code, we can do it" and switch from a bare function to a func with variadic parameters, because we can relatively quickly find out if that breaks anyone (and work with them to deal with the breakage).

Reply:
@kylelemons I'm more concerned about people wanting to specify the port, not the host.

If you really don't think we will want to add any other configuration or customization, then I can implement your suggestions to remove the `ServerArgs`, but I feel its likely people will want to add more endpoints in the future. And the endpoints could need a custom handler.